### PR TITLE
Add optimization over covariance structure to GaussianCluster. Tests …

### PR DESCRIPTION
**Pull request by hhelm10**
…included.

### CODE REVIEW

**Comment by hhelm10** [comment]
@j1c can you please review?

**Comment by j1c** [comment]
1. Do this so your commit is linked to your account: https://help.github.com/en/articles/setting-your-commit-email-address-in-git

2. There is a lot of duplicate code that I think can be simplified. As an extension, we could also allow for different choices (eg. only do `full` and `tied` if someone feels like it). This can be easily achieved via sklearn `ParameterGrid`. Something like below:

```
from sklearn.model_selection import ParameterGrid
from sklearn.mixture import GaussianMixture

param_grid = dict(
    covariance_type = ['full', 'tied', 'diag', 'spherical'], #this list can contain any choices of cov type
    n_components = range(1, max_components),
    random_state = [random_state]
)

param_grid = list(ParameterGrid(param_grid))

for params in param_grid:
    gmm = GaussianMixture(**params)
    #do things
```

**Comment by j1c** [comment]
1. Can you update the docstring according to the changes? 
2. Can you use `black` for formatting?

File: graspy/cluster/gclust.py [hunk 1]
```diff
@@ -56,10 +67,16 @@ class GaussianCluster(BaseCluster):
         all possible number of clusters given by range(1, max_components).
     """
 
-    def __init__(self, max_components=1, covariance_type="full", random_state=None):
+    def __init__(
+        self,
+        max_components=1,
+        min_components=1,
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Switch order of min and max_components. I think it makes sense to go smaller to larger.

File: graspy/cluster/gclust.py [hunk 2]
```diff
@@ -68,7 +85,21 @@ def __init__(self, max_components=1, covariance_type="full", random_state=None):
...
             raise TypeError(msg)
+
+        if isinstance(min_components, int):
+            if min_components <= 0:
+                msg = "min_components must be >= 1."
+                raise ValueError(msg)
+            elif min_components >= max_components:
+                msg = "min_components must be strictly smaller than max_components"
+        else:
+            msg = "max_components must be an integer, not {}.".format(
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Change to `min_components`

**Comment by j1c** [comment]
Change attribute description for `bic_` and `ari_`

File: graspy/cluster/gclust.py [hunk 3]
```diff
@@ -1,29 +1,32 @@
...
 from sklearn.utils.validation import check_is_fitted
+from sklearn.model_selection import ParameterGrid
 
 from .base import BaseCluster
 
 
 class GaussianCluster(BaseCluster):
     r"""
     Gaussian Mixture Model (GMM)
-
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Add a newline back here.

File: graspy/cluster/gclust.py [hunk 4]
```diff
@@ -1,29 +1,32 @@
...
     Representation of a Gaussian mixture model probability distribution. 
     This class allows to estimate the parameters of a Gaussian mixture 
     distribution. It computes all possible models from one component to 
     max_components. The best model is given by the lowest BIC score.
-
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Add a newline here.

File: graspy/cluster/gclust.py [hunk 5]
```diff
@@ -1,29 +1,32 @@
...
+        max_components=None, in which case the parameter is one the highest
+        number of mixture components to consider and must be less than
+        max_components).
     max_components : int, defaults to 1.
         The maximum number of mixture components to consider.
-
     covariance_type : {'full' (default), 'tied', 'diag', 'spherical'}, optional
-        String describing the type of covariance parameters to use.
-        Must be one of:
-
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Add a newline here.

File: graspy/cluster/gclust.py [hunk 6]
```diff
@@ -43,42 +50,85 @@ class GaussianCluster(BaseCluster):
...
+            msg += "string, not {}".format(type(covariance_type))
+            raise TypeError(msg)
+
+        for cov in covariance_type:
+            if cov not in ["spherical", "diag", "tied", "full"]:
+                msg = 'covariance structure must be one of ["spherical", "diag", "tied", "full"]'
+                msg += " not {}".format(cov)
+                raise ValueError(msg)
+
+        new_covariance_type = []
```
Comment (j1c) [inline graspy/cluster/gclust.py]: I think these lines (116-119) are unnecessary.

File: graspy/cluster/gclust.py [hunk 7]
```diff
@@ -88,49 +138,87 @@ def fit(self, X, y=None):
         y : array-like, shape (n_samples,), optional (default=None)
             List of labels for X if available. Used to compute
             ARI scores.
-
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Add newline.

File: graspy/cluster/gclust.py [hunk 8]
```diff
@@ -43,42 +50,85 @@ class GaussianCluster(BaseCluster):
...
+        self.covariance_type = new_covariance_type
         self.random_state = random_state
 
     def fit(self, X, y=None):
         """
         Fits gaussian mixure model to the data. 
-
-
         Estimate model parameters with the EM algorithm.
-
```
Comment (j1c) [inline graspy/cluster/gclust.py]: Add newline

File: graspy/cluster/gclust.py (hunk 5 above)
Comment (hhelm10) [reply graspy/cluster/gclust.py]: is this to separate required parameters from optional ones?

Comment (hhelm10) [reply graspy/cluster/gclust.py]: just a bit confused as to where to add it.

Comment (j1c) [reply graspy/cluster/gclust.py]: After `If a string, it must be one of:` and before `- 'full'`

**Pull request merged**